File: test/data/source/sensors/garmin/garmin_protobuf_test.dart
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';

import 'package:openvitals/data/source/sensors/garmin/garmin_gfdi_frame.dart';
import 'package:openvitals/data/source/sensors/garmin/garmin_messages.dart';
import 'package:openvitals/data/source/sensors/garmin/garmin_protobuf.dart';
import 'package:openvitals/data/source/sensors/garmin/garmin_protobuf_transport.dart';

Uint8List _b(List<int> xs) => Uint8List.fromList(xs);

/// Wraps [payload] the way the watch does, so the transport is exercised
/// against the real envelope rather than a mock of itself.
GarminGfdiFrame _reply(int requestId, Uint8List payload, {int? offset, int? total}) {
  final b = BytesBuilder()
    ..add([requestId & 0xFF, (requestId >> 8) & 0xFF])
    ..add(_u32(offset ?? 0))
    ..add(_u32(total ?? payload.length))
    ..add(_u32(payload.length))
    ..add(payload);
  return GarminGfdiFrame.parse(
    GarminGfdiFrame.build(GarminMessageId.protobufResponse, b.takeBytes()),
  );
}

List<int> _u32(int v) =>
    [v & 0xFF, (v >> 8) & 0xFF, (v >> 16) & 0xFF, (v >> 24) & 0xFF];

void main() {
  group('protobuf encoding', () {
    test('a varint field encodes key then value', () {
      expect((ProtobufWriter()..varint(1, 60)).toBytes(), _b([0x08, 0x3C]));
    });

    test('a multi-byte varint is little-endian base-128', () {
      expect((ProtobufWriter()..varint(1, 300)).toBytes(), _b([0x08, 0xAC, 0x02]));
    });

    test('an empty nested message is not the same as an absent one', () {
      // Garmin uses the empty message as the whole request for actions taking
      // no arguments — cancelling a find is exactly that.
      expect((ProtobufWriter()..emptyMessage(3)).toBytes(), _b([0x1A, 0x00]));
      expect((ProtobufWriter()).toBytes(), isEmpty);
    });

    test('round-trips through the reader', () {
      final inner = (ProtobufWriter()..varint(1, 60)).toBytes();
      final outer = (ProtobufWriter()..nested(12, inner)).toBytes();
      final fields = readProtobuf(outer);
      expect(fields.single.field, 12);
      expect(readProtobuf(fields.single.bytes!).single.varint, 60);
    });

    test('a truncated message yields what was readable, not a crash', () {
      expect(readProtobuf(_b([0x0A, 0x05, 0x01])), isEmpty);
    });
  });

  group('find my watch', () {
    test('start carries a 60-second timeout under the find service', () {
      // Smart.find_my_watch_service = 12, FindMyWatchRequest.timeout = 1.
      expect(GarminFindMyWatch.start(), _b([0x62, 0x04, 0x0A, 0x02, 0x08, 0x3C]));
    });

    test('cancel is an empty message, not a missing one', () {
      expect(GarminFindMyWatch.cancel(), _b([0x62, 0x02, 0x1A, 0x00]));
    });

    test('OK is 100 — a zero status is NOT success', () {
      Uint8List reply(int status) {
        final response = (ProtobufWriter()..varint(1, status)).toBytes();
        final service = (ProtobufWriter()..nested(2, response)).toBytes();
        return (ProtobufWriter()..nested(12, service)).toBytes();
      }

      expect(GarminFindMyWatch.outcome(reply(100)), GarminFindOutcome.ok);
      expect(GarminFindMyWatch.outcome(reply(200)), GarminFindOutcome.error);
      expect(GarminFindMyWatch.outcome(reply(0)), GarminFindOutcome.unknown);
    });

    test('an EMPTY response is acceptance — the real watch sends no status', () {
      // Captured from a vívoactive 5: `62 02 12 00` is find_response present
      // with no status field, and the watch was ringing when it sent it.
      expect(
        GarminFindMyWatch.outcome(_b([0x62, 0x02, 0x12, 0x00])),
        GarminFindOutcome.ok,
      );
      // And the cancel it answers with, field 4.
      expect(
        GarminFindMyWatch.outcome(_b([0x62, 0x02, 0x22, 0x00])),
        GarminFindOutcome.ok,
      );
    });

    test('an unreadable reply is UNKNOWN, never a refusal', () {
      // The watch was seen ringing while its reply was being read as a
      // refusal. Only an explicit ERROR means it declined; everything else has
      // to leave the alert stoppable rather than abandoning it.
      expect(GarminFindMyWatch.outcome(null), GarminFindOutcome.unknown);
      expect(GarminFindMyWatch.outcome(Uint8List(0)), GarminFindOutcome.unknown);
      // Service present, but no response field inside it.
      expect(GarminFindMyWatch.outcome(_b([0x62, 0x00])), GarminFindOutcome.unknown);
      expect(GarminFindMyWatch.outcome(_b([0xFF, 0xFF])), GarminFindOutcome.unknown);
      for (final o in [GarminFindOutcome.ok, GarminFindOutcome.unknown]) {
        expect(o.declined, isFalse);
      }
      expect(GarminFindOutcome.error.declined, isTrue);
    });
  });

  group('protobuf transport', () {
    test('matches a reply to its request by id', () async {
      final sent = <GarminGfdiFrame>[];
      late final GarminProtobufTransport transport;
      transport = GarminProtobufTransport(send: (frame) async {
        final parsed = GarminGfdiFrame.parse(frame);
        sent.add(parsed);
        final requestId = parsed.payload[0] | (parsed.payload[1] << 8);
        transport.handleInbound(_reply(requestId, _b([0x62, 0x00])));
      });

      final reply = await transport.request(GarminFindMyWatch.start());
      expect(reply, _b([0x62, 0x00]));
      expect(sent.single.messageType, GarminMessageId.protobufRequest);
    });

    test('request ids advance, so two requests cannot be confused', () async {
      final ids = <int>[];
      late final GarminProtobufTransport transport;
      transport = GarminProtobufTransport(send: (frame) async {
        final parsed = GarminGfdiFrame.parse(frame);
        final id = parsed.payload[0] | (parsed.payload[1] << 8);
        ids.add(id);
        transport.handleInbound(_reply(id, _b([0x00])));
      });

      await transport.request(GarminFindMyWatch.start());
      await transport.request(GarminFindMyWatch.cancel());
      expect(ids, [1, 2]);
    });

    test('a reply for an unknown id is consumed, not mistaken for ours', () {
      final transport = GarminProtobufTransport(send: (_) async {});
      // The watch starts conversations of its own; an unmatched id is one of
      // those, and must not crash or resolve somebody else's request.
      expect(transport.handleInbound(_reply(999, _b([0x01]))), isTrue);
    });

    test('reassembles a chunked reply', () async {
      late final GarminProtobufTransport transport;
      transport = GarminProtobufTransport(send: (frame) async {
        final id = GarminGfdiFrame.parse(frame).payload[0];
        transport.handleInbound(_reply(id, _b([1, 2, 3]), offset: 0, total: 6));
        transport.handleInbound(_reply(id, _b([4, 5, 6]), offset: 3, total: 6));
      });

      expect(await transport.request(GarminFindMyWatch.start()),
          _b([1, 2, 3, 4, 5, 6]));
    });

    test('a dropped link fails the request instead of hanging on it', () async {
      late final GarminProtobufTransport transport;
      transport = GarminProtobufTransport(send: (_) async {});
      final pending = transport.request(GarminFindMyWatch.start());
      transport.abort();
      await expectLater(pending, throwsA(isA<StateError>()));
    });

    test('an oversized payload is refused rather than truncated', () {
      final transport = GarminProtobufTransport(send: (_) async {});
      expect(
        () => transport.request(Uint8List(400)),
        throwsA(isA<ArgumentError>()),
      );
    });
  });
}

File: app/src/test/kotlin/tech/mmarca/openvitals/devices/garmin/GarminProtobufTest.kt
package tech.mmarca.openvitals.devices.garmin

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Test
import kotlinx.coroutines.launch
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.yield

/** Port of the Flutter build's `garmin_protobuf_test.dart` — fixtures identical. */
class GarminProtobufTest {

    private fun b(vararg xs: Int) = ByteArray(xs.size) { xs[it].toByte() }

    /** Wraps [payload] the way the watch does, so the transport sees the real envelope. */
    private fun reply(
        requestId: Int,
        payload: ByteArray,
        offset: Int? = null,
        total: Int? = null,
    ): GarminGfdiFrame {
        val writer = GarminByteWriter()
            .writeShort(requestId)
            .writeInt(offset ?: 0)
            .writeInt(total ?: payload.size)
            .writeInt(payload.size)
            .writeBytes(payload)
        return GarminGfdiFrame.parse(
            GarminGfdiFrame.build(GarminMessageId.PROTOBUF_RESPONSE, writer.toBytes()),
        )
    }

    private fun requestChunk(
        id: Int,
        payload: ByteArray,
        offset: Int,
        total: Int,
    ): GarminGfdiFrame {
        val writer = GarminByteWriter()
            .writeShort(id)
            .writeInt(offset)
            .writeInt(total)
            .writeInt(payload.size)
            .writeBytes(payload)
        return GarminGfdiFrame.parse(
            GarminGfdiFrame.build(GarminMessageId.PROTOBUF_REQUEST, writer.toBytes()),
        )
    }

    // Protobuf encoding.

    @Test
    fun `a varint field encodes key then value`() {
        assertArrayEquals(b(0x08, 0x3C), ProtobufWriter().varint(1, 60).toBytes())
    }

    @Test
    fun `a multi-byte varint is little-endian base-128`() {
        assertArrayEquals(b(0x08, 0xAC, 0x02), ProtobufWriter().varint(1, 300).toBytes())
    }

    @Test
    fun `an empty nested message is not the same as an absent one`() {
        // Garmin uses the empty message as the whole request for actions taking no arguments.
        assertArrayEquals(b(0x1A, 0x00), ProtobufWriter().emptyMessage(3).toBytes())
        assertEquals(0, ProtobufWriter().toBytes().size)
    }

    @Test
    fun `round-trips through the reader`() {
        val inner = ProtobufWriter().varint(1, 60).toBytes()
        val outer = ProtobufWriter().nested(12, inner).toBytes()
        val fields = readProtobuf(outer)
        assertEquals(12, fields.single().field)
        assertEquals(60L, readProtobuf(fields.single().bytes!!).single().varint)
    }

    @Test
    fun `a truncated message yields what was readable not a crash`() {
        assertTrue(readProtobuf(b(0x0A, 0x05, 0x01)).isEmpty())
    }

    // Find my watch.

    @Test
    fun `start carries a 60-second timeout under the find service`() {
        // Smart.find_my_watch_service = 12, FindMyWatchRequest.timeout = 1.
        assertArrayEquals(
            b(0x62, 0x04, 0x0A, 0x02, 0x08, 0x3C),
            GarminFindMyWatch.start(),
        )
    }

    @Test
    fun `cancel is an empty message not a missing one`() {
        assertArrayEquals(b(0x62, 0x02, 0x1A, 0x00), GarminFindMyWatch.cancel())
    }

    @Test
    fun `OK is 100 — a zero status is NOT success`() {
        fun replyWithStatus(status: Int): ByteArray {
            val response = ProtobufWriter().varint(1, status).toBytes()
            val service = ProtobufWriter().nested(2, response).toBytes()
            return ProtobufWriter().nested(12, service).toBytes()
        }

        assertEquals(GarminFindOutcome.OK, GarminFindMyWatch.outcome(replyWithStatus(100)))
        assertEquals(GarminFindOutcome.ERROR, GarminFindMyWatch.outcome(replyWithStatus(200)))
        assertEquals(GarminFindOutcome.UNKNOWN, GarminFindMyWatch.outcome(replyWithStatus(0)))
    }

    @Test
    fun `an EMPTY response is acceptance — the real watch sends no status`() {
        // From a vívoactive 5: `62 02 12 00` is find_response with no status, and the watch was ringing.
        assertEquals(
            GarminFindOutcome.OK,
            GarminFindMyWatch.outcome(b(0x62, 0x02, 0x12, 0x00)),
        )
        // And the cancel it answers with, field 4.
        assertEquals(
            GarminFindOutcome.OK,
            GarminFindMyWatch.outcome(b(0x62, 0x02, 0x22, 0x00)),
        )
    }

    @Test
    fun `an unreadable reply is UNKNOWN never a refusal`() {
        // Only an explicit ERROR means the watch declined; everything else leaves the alert stoppable.
        assertEquals(GarminFindOutcome.UNKNOWN, GarminFindMyWatch.outcome(null))
        assertEquals(GarminFindOutcome.UNKNOWN, GarminFindMyWatch.outcome(ByteArray(0)))
        // Service present, but no response field inside it.
        assertEquals(GarminFindOutcome.UNKNOWN, GarminFindMyWatch.outcome(b(0x62, 0x00)))
        assertEquals(GarminFindOutcome.UNKNOWN, GarminFindMyWatch.outcome(b(0xFF, 0xFF)))
        for (outcome in listOf(GarminFindOutcome.OK, GarminFindOutcome.UNKNOWN)) {
            assertFalse(outcome.declined)
        }
        assertTrue(GarminFindOutcome.ERROR.declined)
    }

    // Protobuf transport.

    @Test
    fun `matches a reply to its request by id`() = runTest {
        val sent = mutableListOf<GarminGfdiFrame>()
        lateinit var transport: GarminProtobufTransport
        transport = GarminProtobufTransport(send = { frame ->
            val parsed = GarminGfdiFrame.parse(frame)
            // Answer the request only; replying to the transport's own acknowledgements recurses.
            if (parsed.messageType != GarminMessageId.PROTOBUF_REQUEST) return@GarminProtobufTransport
            sent.add(parsed)
            val requestId = (parsed.payload[0].toInt() and 0xFF) or
                ((parsed.payload[1].toInt() and 0xFF) shl 8)
            transport.handleInbound(reply(requestId, b(0x62, 0x00)))
        })

        val result = transport.request(GarminFindMyWatch.start())
        assertArrayEquals(b(0x62, 0x00), result)
        assertEquals(GarminMessageId.PROTOBUF_REQUEST, sent.single().messageType)
    }

    @Test
    fun `request ids advance so two requests cannot be confused`() = runTest {
        val ids = mutableListOf<Int>()
        lateinit var transport: GarminProtobufTransport
        transport = GarminProtobufTransport(send = { frame ->
            val parsed = GarminGfdiFrame.parse(frame)
            if (parsed.messageType != GarminMessageId.PROTOBUF_REQUEST) return@GarminProtobufTransport
            val id = (parsed.payload[0].toInt() and 0xFF) or
                ((parsed.payload[1].toInt() and 0xFF) shl 8)
            ids.add(id)
            transport.handleInbound(reply(id, b(0x00)))
        })

        transport.request(GarminFindMyWatch.start())
        transport.request(GarminFindMyWatch.cancel())
        assertEquals(listOf(1, 2), ids)
    }

    @Test
    fun `a COMPLETE message is acknowledged by request id not generically`() = runTest {
        // The watch also wants to hear the protobuf message was kept. Without that it retransmitted
        // every message every five seconds, so a stale reply was in flight during a different request.
        val acks = mutableListOf<GarminGfdiFrame>()
        val transport = GarminProtobufTransport(send = { frame ->
            val parsed = GarminGfdiFrame.parse(frame)
            if (parsed.messageType == GarminMessageId.RESPONSE) acks.add(parsed)
        })

        transport.handleInbound(reply(4242, b(0x62, 0x00)))

        val ack = acks.single().payload
        // [u16 acked type][u8 ACK][u16 requestId][u32 offset][kept][no error]
        assertEquals(11, ack.size)
        assertEquals(4242, (ack[3].toInt() and 0xFF) or ((ack[4].toInt() and 0xFF) shl 8))
        assertArrayEquals(b(0, 0, 0, 0), ack.copyOfRange(5, 9))
        assertArrayEquals(b(0, 0), ack.copyOfRange(9, 11))
    }

    @Test
    fun `a reply for an unknown id is consumed not mistaken for ours`() = runTest {
        val transport = GarminProtobufTransport(send = { })
        // The watch starts conversations of its own; an unmatched id must not resolve somebody else's request.
        assertTrue(transport.handleInbound(reply(999, b(0x01))))
    }

    @Test
    fun `reassembles a chunked reply`() = runTest {
        lateinit var transport: GarminProtobufTransport
        transport = GarminProtobufTransport(send = { frame ->
            val parsed = GarminGfdiFrame.parse(frame)
            // Answer the request only; replying to chunk acknowledgements recurses.
            if (parsed.messageType != GarminMessageId.PROTOBUF_REQUEST) return@GarminProtobufTransport
            val id = parsed.payload[0].toInt() and 0xFF
            transport.handleInbound(reply(id, b(1, 2, 3), offset = 0, total = 6))
            transport.handleInbound(reply(id, b(4, 5, 6), offset = 3, total = 6))
        })

        assertArrayEquals(
            b(1, 2, 3, 4, 5, 6),
            transport.request(GarminFindMyWatch.start()),
        )
    }

    @Test
    fun `a dropped link fails the request instead of hanging on it`() = runTest {
        val transport = GarminProtobufTransport(send = { })
        var failure: Throwable? = null
        val job = launch {
            try {
                transport.request(GarminFindMyWatch.start())
            } catch (error: IllegalStateException) {
                failure = error
            }
        }
        yield() // Let the request register itself.
        transport.abort()
        job.join()
        assertTrue(failure is IllegalStateException)
    }

    @Test
    fun `an oversized payload is refused rather than truncated`() = runTest {
        val transport = GarminProtobufTransport(send = { })
        try {
            transport.request(ByteArray(400))
            fail("expected IllegalArgumentException")
        } catch (expected: IllegalArgumentException) {
            // Expected.
        }
    }

    // Unsolicited chunking.

    @Test
    fun `reassembles a message the watch sent under its OWN id`() = runTest {
        // The watch answers a settings request with its own id, so accumulation cannot be keyed on waiting.
        val delivered = mutableListOf<ByteArray>()
        val acks = mutableListOf<GarminGfdiFrame>()
        val transport = GarminProtobufTransport(
            send = { frame -> acks.add(GarminGfdiFrame.parse(frame)) },
            onUnsolicited = { delivered.add(it) },
        )

        transport.handleInbound(requestChunk(324, b(1, 2, 3), 0, 6))
        assertTrue("incomplete must not be delivered", delivered.isEmpty())
        transport.handleInbound(requestChunk(324, b(4, 5, 6), 3, 6))

        assertArrayEquals(b(1, 2, 3, 4, 5, 6), delivered.single())
    }

    @Test
    fun `acknowledges a chunk with the offset IT declared`() = runTest {
        // Not the next offset. Echoing `dataOffset + chunkLength` made the watch resend chunk zero forever.
        val acks = mutableListOf<GarminGfdiFrame>()
        val transport = GarminProtobufTransport(
            send = { frame -> acks.add(GarminGfdiFrame.parse(frame)) },
            onUnsolicited = { },
        )

        transport.handleInbound(requestChunk(324, b(4, 5, 6), 3, 6))

        val payload = acks.single().payload
        assertEquals(GarminMessageId.RESPONSE, acks.single().messageType)
        // [u16 originalType][u8 status][u16 requestId][u32 dataOffset][u8][u8]
        val requestId = (payload[3].toInt() and 0xFF) or ((payload[4].toInt() and 0xFF) shl 8)
        val offset = (payload[5].toInt() and 0xFF) or
            ((payload[6].toInt() and 0xFF) shl 8) or
            ((payload[7].toInt() and 0xFF) shl 16) or
            ((payload[8].toInt() and 0xFF) shl 24)
        assertEquals(324, requestId)
        assertEquals("the offset received, not the next expected", 3, offset)
    }
}
